mobilecoind-dev-faucet/src/slam/mod.rs

```diff
@@ -0,0 +1,500 @@
+}
+
+impl Default for SlamParams {
+    fn default() -> Self {
+        Self {
+            target_num_tx: 500,
+            num_threads: 30,
+            retries: 30,
+            retry_period: Duration::from_millis(1000),
+            tombstone_offset: 10,
+            consensus_client_uris: Default::default(),
+        }
+    }
+}
+
+/// A report on a completed slam operation
+#[derive(Clone, Debug, Default)]
+pub struct SlamReport {
+    /// The number of utxos successfully prepared
+    pub num_prepared_utxos: usize,
+    /// The number of txs successfully submitted to the network
+    pub num_submitted_txs: usize,
+    /// The total time spent preparing Txs
+    pub prepare_time: Duration,
+    /// The total time spent submitting Txs
+    pub submit_time: Duration,
+}
+
+/// State which tracks the progress of a running slam operation.
+/// This allows one thread to conduct the slam while another thread checks on
+/// its status.
+pub struct SlamState {
+    // Phase is a state variable:
+    // 0: Not currently slamming
+    // 1: Connecting
+    // 2: Preparing utxos: Obtaining proofs of membership
+    // 3: Submitting transactions
+    phase: AtomicUsize,
+    // The number of tx's we want to send in total in the slam
+    target_num_tx: AtomicUsize,
+    // The number of utxo's we have prepared so far in this slam
+    num_prepared_utxos: AtomicUsize,
+    // The number of tx's we have successfully submitted so far in this slam
+    num_submitted_txs: AtomicUsize,
+    // The estimated block height of the network, updated during slam as we submit to consensus
+    block_height: AtomicU64,
+    // Whether it has been requested to stop the slam
+    stop_requested: AtomicBool,
+    // grpcio environment
+    env: Arc<grpcio::Environment>,
+}
+
+impl SlamState {
+    /// Create a new slam state
+    pub fn new(env: Arc<grpcio::Environment>) -> Self {
+        Self {
+            phase: Default::default(),
+            target_num_tx: Default::default(),
+            num_prepared_utxos: Default::default(),
+            num_submitted_txs: Default::default(),
+            block_height: Default::default(),
+            stop_requested: Default::default(),
+            env,
+        }
+    }
+
+    /// Start a slam operation with given parameters
+    ///
+    /// This returns immediately with an error if another slam is currently in
+    /// progress, otherwise it starts a new slam.
+    /// Other threads can track the progress of the slam by calling
+    /// "get_status()".
+    ///
+    /// The slam has three steps:
+    /// (1) connecting to consensus
+    /// (2) getting rings and proofs of membership for utxos
+    /// (3) building and submitting Tx's in parallel
+    ///
+    /// This function returns a slam report, which includes how many Txs were
+    /// ultimately submitted successfully and how long each step took, or an
+    /// error message.
+    pub async fn start_slam(
+        self: Arc<Self>,
+        params: SlamParams,
```

Comment:
sure

Reply:
I'm getting an interesting error when I try to change this:

```
error[E0505]: cannot move out of `params` because it is borrowed
   --> mobilecoind-dev-faucet/src/lib.rs:423:27
    |
401 |             &params,
    |             ------- borrow of `params` occurs here
...
423 |         Ok(SlamResponse { params, report })
    |                           ^^^^^^ move out of `params` occurs here
424 |     }
    |     - borrow might be used here, when `slam_fut` is dropped and runs the destructor for type `Pin<Box<impl Future<Output = Result<SlamReport, std::string::String>>>>`

```

I can resolve it by manually dropping `slam_fut` but it seems inelegant